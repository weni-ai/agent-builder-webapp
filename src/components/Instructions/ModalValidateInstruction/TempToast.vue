<!-- 
 This temporary component should be replaced by Unnnic system's Toast as soon as
 it is in a stable version and includes the static variation.
 -->
<template>
  <section :class="[`temp-toast temp-toast--${type}`]">
    <UnnnicIcon
      :icon="type === 'warning' ? 'error' : 'check_circle'"
      :scheme="type === 'warning' ? 'yellow-500' : 'green-500'"
      size="ant"
    />
    <h3 class="temp-toast__title">
      {{ title }}
    </h3>
    <p
      v-if="description"
      class="temp-toast__description"
    >
      {{ description }}
    </p>
  </section>
</template>

<script setup lang="ts">
defineProps<{
  type: 'warning' | 'success';
  title: string;
  description?: string;
}>();
</script>

<style lang="scss" scoped>
.temp-toast {
  border-radius: $unnnic-radius-2;

  padding: $unnnic-space-4;

  display: grid;
  grid-template-columns: auto 1fr;
  align-items: center;
  gap: $unnnic-space-2;

  &--warning {
    border: 1px solid $unnnic-color-border-warning;
    background-color: $unnnic-color-bg-warning;
  }

  &--success {
    border: 1px solid $unnnic-color-border-success;
    background-color: $unnnic-color-bg-success;
  }

  &__title {
    font: $unnnic-font-action;
    color: $unnnic-color-fg-emphasized;
  }

  &__description {
    grid-column: 2 / 3;

    font: $unnnic-font-caption-2;
    color: $unnnic-color-fg-emphasized;
  }
}
</style>
