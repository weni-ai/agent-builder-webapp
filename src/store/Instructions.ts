import { reactive, ref, computed } from 'vue';
import { defineStore } from 'pinia';

import type { InstructionSuggestedByAI } from './types/Instructions.types';

import { useProjectStore } from './Project';
import { useAlertStore } from './Alert';

import nexusaiAPI from '@/api/nexusaiAPI';

import i18n from '@/utils/plugins/i18n';
import { moduleStorage } from '@/utils/storage';

function callAlert(type, alertText) {
  const alertStore = useAlertStore();
  alertStore.add({
    text: i18n.global.t(`agent_builder.instructions.${alertText}`),
    type,
  });
}

export const useInstructionsStore = defineStore('Instructions', () => {
  const projectUuid = computed(() => useProjectStore().uuid);

  const instructions = reactive({
    data: [],
    status: null,
  });

  const newInstruction = reactive({
    text: '',
    status: null,
  });

  const validateInstructionByAI = ref<boolean>(
    moduleStorage.getItem('validateInstructionByAI') ?? true,
  );

  const instructionSuggestedByAI = reactive<InstructionSuggestedByAI>({
    data: {
      instruction: '',
      classification: [],
      suggestion: '',
    },
    suggestionApplied: false,
    status: null,
  });

  const activeInstructionsListTab = ref('custom');

  async function addInstruction() {
    newInstruction.status = 'loading';
    activeInstructionsListTab.value = 'custom';

    try {
      const instructionResponse =
        await nexusaiAPI.agent_builder.instructions.addInstruction({
          projectUuid: projectUuid.value,
          instruction: newInstruction,
        });

      instructions.data.unshift({
        ...newInstruction,
        status: 'complete',
        id: instructionResponse.id,
      });
      newInstruction.status = null;
      newInstruction.text = '';

      callAlert('success', 'new_instruction.success_alert');
    } catch (error) {
      newInstruction.status = 'error';
      callAlert('error', 'new_instruction.error_alert');
    }
  }

  async function loadInstructions() {
    instructions.status = 'loading';
    try {
      const response = await nexusaiAPI.agent_builder.instructions.list({
        projectUuid: projectUuid.value,
      });

      instructions.data = [...instructions.data, ...response];
      instructions.status = 'complete';
    } catch (error) {
      instructions.status = 'error';
    }
  }
  async function editInstruction(id, text) {
    const instruction = instructions.data.find(
      (instruction) => instruction.id === id,
    );

    if (!instruction) return;

    try {
      instruction.status = 'loading';
      await nexusaiAPI.agent_builder.instructions.edit({
        projectUuid: projectUuid.value,
        id,
        text,
      });
      instruction.text = text;
      instruction.status = 'complete';

      callAlert('success', 'edit_instruction.success_alert');
    } catch (error) {
      instruction.status = 'error';
      callAlert('error', 'edit_instruction.error_alert');
    }

    return { status: instruction.status };
  }

  async function removeInstruction(id) {
    const instruction = instructions.data.find(
      (instruction) => instruction.id === id,
    );
    if (!instruction) return;
    instruction.status = 'loading';

    try {
      await nexusaiAPI.agent_builder.instructions.delete({
        projectUuid: projectUuid.value,
        id,
      });
      instructions.data = instructions.data.filter(
        (instruction) => instruction.id !== id,
      );
      callAlert('default', 'remove_instruction.success_alert');
      return { status: null };
    } catch (error) {
      instruction.status = 'error';
      callAlert('error', 'remove_instruction.error_alert');
    }

    return { status: instruction?.status };
  }

  async function getInstructionSuggestionByAI() {
    instructionSuggestedByAI.status = 'loading';

    try {
      const response =
        await nexusaiAPI.agent_builder.instructions.getSuggestionByAI({
          projectUuid: projectUuid.value,
          instruction: newInstruction.text,
        });

      instructionSuggestedByAI.data = {
        instruction: newInstruction.text,
        ...response,
      };
      instructionSuggestedByAI.suggestionApplied = false;
      instructionSuggestedByAI.status = 'complete';
    } catch (error) {
      instructionSuggestedByAI.status = 'error';
      callAlert(
        'error',
        'new_instruction.validate_instruction_by_ai.error_alert',
      );
    }
  }

  function updateValidateInstructionByAI(value: boolean) {
    validateInstructionByAI.value = value;
    moduleStorage.setItem('validateInstructionByAI', value);
  }

  function resetInstructionSuggestedByAI() {
    instructionSuggestedByAI.data = {
      instruction: '',
      classification: [],
      suggestion: '',
    };
  }

  return {
    instructions,
    newInstruction,
    validateInstructionByAI,
    instructionSuggestedByAI,
    activeInstructionsListTab,
    addInstruction,
    loadInstructions,
    editInstruction,
    removeInstruction,
    getInstructionSuggestionByAI,
    updateValidateInstructionByAI,
    resetInstructionSuggestedByAI,
  };
});
