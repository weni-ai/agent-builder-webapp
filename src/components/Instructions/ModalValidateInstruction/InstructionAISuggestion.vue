<template>
  <UnnnicPopover>
    <UnnnicPopoverTrigger class="instruction-ai-suggestion__trigger">
      <UnnnicIcon
        icon="lightbulb"
        size="sm"
        scheme="teal-600"
      />
      {{ aiSuggestionTranslation('trigger') }}
    </UnnnicPopoverTrigger>
    <UnnnicPopoverContent
      align="end"
      size="large"
      class="instruction-ai-suggestion__content"
    >
      <h3 class="instruction-ai-suggestion__title">
        {{ aiSuggestionTranslation('title') }}
      </h3>
      <UnnnicTextArea
        :modelValue="aiSuggestion"
        :placeholder="
          $t('agent_builder.instructions.new_instruction.textarea.placeholder')
        "
        class="instruction-ai-suggestion__textarea"
      />
      <template #footer>
        <UnnnicButton>
          {{ aiSuggestionTranslation('apply_button') }}
        </UnnnicButton>
      </template>
    </UnnnicPopoverContent>
  </UnnnicPopover>
</template>

<script setup lang="ts">
import { useInstructionsStore } from '@/store/Instructions';
import { computed } from 'vue';
import i18n from '@/utils/plugins/i18n';

const instructionsStore = useInstructionsStore();

const aiSuggestionTranslation = (key: string) =>
  i18n.global.t(
    `agent_builder.instructions.new_instruction.validate_instruction_by_ai.see_ai_suggestion.${key}`,
  );

const aiSuggestion = computed(() => {
  return instructionsStore.instructionSuggestedByAI.data.suggestion;
});
</script>

<style lang="scss">
.instruction-ai-suggestion__trigger {
  position: absolute;
  right: $unnnic-space-4;
  bottom: $unnnic-space-3;

  font: $unnnic-font-action;
  color: $unnnic-color-fg-active;
}

.instruction-ai-suggestion__content {
  display: grid;
  gap: $unnnic-space-4;
}

.instruction-ai-suggestion__title {
  font: $unnnic-font-display-3;
  color: $unnnic-color-fg-emphasized;
}

.instruction-ai-suggestion__textarea {
  .unnnic-text-area__textarea {
    resize: none;
  }
}
</style>
