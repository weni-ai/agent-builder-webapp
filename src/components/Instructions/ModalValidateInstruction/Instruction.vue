<template>
  <section
    data-testid="modal-validate-instruction"
    class="modal-validate-instruction"
  >
    <section class="modal-validate-instruction__instruction">
      <UnnnicTextArea
        v-model="modelValue"
        data-testid="instruction-textarea"
        :placeholder="
          $t('agent_builder.instructions.new_instruction.textarea.placeholder')
        "
      />
      <InstructionAISuggestion
        v-if="!isSuggestionLoading"
        @apply-suggestion="applySuggestion"
      />
    </section>
    <UnnnicButton
      data-testid="revalidate-button"
      :text="
        $t(
          'agent_builder.instructions.new_instruction.validate_instruction_by_ai.re-validate_button',
        )
      "
      :disabled="revalidateButtonTextDisabled"
      :loading="isSuggestionLoading"
      @click="revalidateInstructionByAI"
    />
  </section>
</template>

<script setup lang="ts">
import { computed } from 'vue';

import { useInstructionsStore } from '@/store/Instructions';

import InstructionAISuggestion from './InstructionAISuggestion.vue';

const instructionsStore = useInstructionsStore();

const modelValue = defineModel<string>('modelValue');

const isSuggestionLoading = computed(() => {
  return instructionsStore.instructionSuggestedByAI.status === 'loading';
});

const revalidateButtonTextDisabled = computed(() => {
  const isInstructionEqualsToNewInstruction =
    modelValue.value?.trim() === instructionsStore.newInstruction.text?.trim();

  return (
    !modelValue.value?.trim() ||
    isInstructionEqualsToNewInstruction ||
    instructionsStore.instructionSuggestedByAI.suggestionApplied
  );
});

function revalidateInstructionByAI() {
  instructionsStore.getInstructionSuggestionByAI();
}

function applySuggestion(suggestion: string) {
  modelValue.value = suggestion;
  instructionsStore.instructionSuggestedByAI.suggestionApplied = true;
  instructionsStore.resetInstructionSuggestedByAI();
}
</script>

<style lang="scss" scoped>
.modal-validate-instruction {
  display: grid;
  gap: $unnnic-space-2;

  &__instruction {
    position: relative;
  }
}
</style>
