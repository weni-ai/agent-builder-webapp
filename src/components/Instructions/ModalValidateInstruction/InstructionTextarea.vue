<template>
  <section class="modal-validate-instruction__textarea">
    <UnnnicTextArea
      v-model="modelValue"
      :placeholder="
        $t('agent_builder.instructions.new_instruction.textarea.placeholder')
      "
      resize="none"
    />
    <UnnnicButton
      :text="
        $t(
          'agent_builder.instructions.new_instruction.validate_instruction_by_ai.re-validate_button',
        )
      "
      :disabled="revalidateButtonTextDisabled"
      :loading="instructionsStore.instructionSuggestedByAI.status === 'loading'"
      @click="revalidateInstructionByAI"
    />
  </section>
</template>

<script setup lang="ts">
import { computed } from 'vue';

import { useInstructionsStore } from '@/store/Instructions';

const instructionsStore = useInstructionsStore();

const modelValue = defineModel<string>('modelValue');

const revalidateButtonTextDisabled = computed(
  () =>
    !modelValue.value?.trim() ||
    modelValue.value?.trim() === instructionsStore.newInstruction.text?.trim(),
);

function revalidateInstructionByAI() {
  instructionsStore.getInstructionSuggestionByAI();
}
</script>

<style lang="scss" scoped>
.modal-validate-instruction__textarea {
  display: grid;
  gap: $unnnic-space-2;
}
</style>
