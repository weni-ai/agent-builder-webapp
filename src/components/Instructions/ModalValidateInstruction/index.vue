<template>
  <UnnnicModalDialog
    data-testid="modal-validate-instruction-by-ai"
    :modelValue="modelValue"
    size="lg"
    showCloseIcon
    showActionsDivider
    :title="
      $t(
        'agent_builder.instructions.new_instruction.validate_instruction_by_ai.modal_title',
      )
    "
    :secondaryButtonProps="{
      text: $t(
        'agent_builder.instructions.new_instruction.validate_instruction_by_ai.cancel_button',
      ),
      onClick: close,
    }"
    :primaryButtonProps="{
      text: $t(
        'agent_builder.instructions.new_instruction.validate_instruction_by_ai.publish_button',
      ),
      disabled:
        instructionsStore.instructionSuggestedByAI.status === 'loading' ||
        newInstruction.trim() === '',
      loading: instructionsStore.newInstruction.status === 'loading',
      onClick: publishInstruction,
    }"
    @update:model-value="close"
  >
    <InstructionTextarea
      v-model="newInstruction"
      data-testid="modal-validate-instruction-textarea"
    />
  </UnnnicModalDialog>
</template>

<script setup lang="ts">
import { onMounted, ref } from 'vue';

import { useInstructionsStore } from '@/store/Instructions';

import InstructionTextarea from './InstructionTextarea.vue';

const emit = defineEmits(['update:modelValue']);

const instructionsStore = useInstructionsStore();

const modelValue = ref(false);

const newInstruction = ref(instructionsStore.newInstruction.text);

function close() {
  emit('update:modelValue', false);
}

function publishInstruction() {
  instructionsStore.newInstruction.text = newInstruction.value;
  instructionsStore.addInstruction();
}

onMounted(() => {
  const isInstructionNotSuggested =
    instructionsStore.instructionSuggestedByAI.data.instruction !==
    instructionsStore.newInstruction.text;

  if (isInstructionNotSuggested) {
    instructionsStore.getInstructionSuggestionByAI();
  }
});
</script>
