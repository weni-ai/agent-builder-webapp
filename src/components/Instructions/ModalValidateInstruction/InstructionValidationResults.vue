<template>
  <section
    data-testid="modal-validate-instruction-validation-results"
    class="modal-validate-instruction__validation-results"
  >
    <h2 class="modal-validate-instruction__title">
      {{
        $t(
          'agent_builder.instructions.new_instruction.validate_instruction_by_ai.results.title',
        )
      }}
    </h2>
    <UnnnicSkeletonLoading
      v-if="instructionsStore.instructionSuggestedByAI.status === 'loading'"
      class="modal-validate-instruction__loading"
      tag="div"
      width="100%"
      height="81px"
    />
    <template
      v-else-if="
        instructionsStore.instructionSuggestedByAI.status === 'complete'
      "
    >
      <TempToast
        v-for="classification in classifications"
        :key="classification.type"
        :type="classification.type"
        :title="classification.name"
        :description="classification.reason"
      />
    </template>
  </section>
</template>

<script setup lang="ts">
import { computed } from 'vue';

import { useInstructionsStore } from '@/store/Instructions';
import { Classification } from '@/store/types/Instructions.types';

import i18n from '@/utils/plugins/i18n';

import TempToast from './TempToast.vue';

const instructionsStore = useInstructionsStore();
const classifications = computed<
  { type: 'warning' | 'success'; name: string; reason: string }[]
>(() => {
  const getTranslation = (key: string) =>
    i18n.global.t(
      `agent_builder.instructions.new_instruction.validate_instruction_by_ai.results.${key}`,
    );

  const titleMappings = {
    duplicate: getTranslation('duplicate'),
    conflict: getTranslation('conflicts'),
    ambiguity: getTranslation('ambiguity'),
    lack_of_clarity: getTranslation('lack_of_clarity'),
  };
  const results =
    instructionsStore.instructionSuggestedByAI.data.classification;

  if (results.length) {
    return results.map((result: Classification) => ({
      ...result,
      type: 'warning' as const,
      name: titleMappings[result.name] || result.reason,
    }));
  }

  return [
    {
      type: 'success' as const,
      name: getTranslation('no_problems_found'),
      reason: '',
    },
  ];
});
</script>

<style lang="scss" scoped>
.modal-validate-instruction {
  &__validation-results {
    display: grid;
    gap: $unnnic-space-4;
  }

  &__title {
    font: $unnnic-font-display-3;
    color: $unnnic-color-fg-emphasized;
  }

  &__loading {
    display: flex;
  }
}
</style>
