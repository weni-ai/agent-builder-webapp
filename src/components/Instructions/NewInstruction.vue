<template>
  <section
    class="new-instruction"
    data-testid="new-instruction"
  >
    <header class="new-instruction__header">
      <h2
        class="new-instruction__title"
        data-testid="new-instruction-title"
      >
        {{ $t('agent_builder.instructions.new_instruction.title') }}
      </h2>
      <UnnnicSwitch
        :modelValue="instructionsStore.validateInstructionByAI"
        :textRight="
          $t(
            'agent_builder.instructions.new_instruction.validate_instruction_by_ai.switch',
          )
        "
        @update:model-value="
          instructionsStore.updateValidateInstructionByAI($event)
        "
      />
    </header>
    <UnnnicTextArea
      v-model="instructionsStore.newInstruction.text"
      data-testid="new-instruction-textarea"
      :placeholder="
        $t('agent_builder.instructions.new_instruction.textarea.placeholder')
      "
      :message="
        $t('agent_builder.instructions.new_instruction.textarea.description')
      "
    />
    <UnnnicButton
      class="new-instruction__add-instruction-button"
      data-testid="add-instruction-button"
      :disabled="!newInstruction.text.trim()"
      :text="primaryButtonText"
      :loading="newInstruction.status === 'loading'"
      @click="handlePrimaryButton"
    />
  </section>
</template>

<script setup>
import { computed } from 'vue';
import i18n from '@/utils/plugins/i18n';
import { useInstructionsStore } from '@/store/Instructions';

const instructionsStore = useInstructionsStore();

const newInstruction = computed(() => instructionsStore.newInstruction);
const primaryButtonText = computed(() => {
  const newInstructionText = (value) =>
    i18n.global.t(`agent_builder.instructions.new_instruction.${value}`);

  return instructionsStore.validateInstructionByAI
    ? newInstructionText('validate_instruction_by_ai.button')
    : newInstructionText('publish_instruction');
});

function openValidateInstructionByAIModal() {
  console.log('openValidateInstructionByAIModal');
}

function handlePrimaryButton() {
  return instructionsStore.validateInstructionByAI
    ? openValidateInstructionByAIModal()
    : instructionsStore.addInstruction();
}
</script>

<style lang="scss" scoped>
.new-instruction {
  border-radius: $unnnic-border-radius-md;
  border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;
  padding: $unnnic-spacing-sm;

  display: grid;
  grid-template-columns: repeat(12, 1fr);
  row-gap: $unnnic-spacing-sm;

  & > * {
    grid-column: 1 / -1;
  }

  &__header {
    display: flex;
    justify-content: space-between;
    align-items: center;
    gap: $unnnic-space-2;
  }

  &__title {
    margin: 0;

    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-lg;
    font-weight: $unnnic-font-weight-bold;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__add-instruction-button {
    grid-column: 1 / 5;
  }
}
</style>
