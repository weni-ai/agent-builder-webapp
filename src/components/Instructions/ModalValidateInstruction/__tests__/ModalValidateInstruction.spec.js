import { mount } from '@vue/test-utils';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { createTestingPinia } from '@pinia/testing';

import ModalValidateInstruction from '../index.vue';
import i18n from '@/utils/plugins/i18n';
import { useInstructionsStore } from '@/store/Instructions';

describe('ModalValidateInstruction.vue', () => {
  let wrapper;
  let instructionsStore;

  const pinia = createTestingPinia({
    initialState: {
      Instructions: {
        newInstruction: {
          text: 'Test instruction text',
          status: null,
        },
        instructionSuggestedByAI: {
          data: {
            instruction: '',
            classification: '',
            reason: '',
            suggestion: '',
          },
          status: null,
        },
      },
    },
  });

  const SELECTORS = {
    modal: '[data-testid="modal-validate-instruction-by-ai"]',
    instructionTextarea: '[data-testid="modal-validate-instruction-textarea"]',
    instructionDivider: '[data-testid="modal-validate-instruction-divider"]',
    instructionValidationResults:
      '[data-testid="modal-validate-instruction-validation-results"]',
  };
  const findComponent = (component) =>
    wrapper.findComponent(SELECTORS[component]);

  beforeEach(() => {
    instructionsStore = useInstructionsStore(pinia);
    instructionsStore.addInstruction = vi.fn().mockResolvedValue({
      status: 'complete',
    });
    instructionsStore.getInstructionSuggestionByAI = vi.fn().mockResolvedValue({
      status: 'complete',
    });

    wrapper = mount(ModalValidateInstruction, {
      props: {
        modelValue: true,
      },
      global: {
        plugins: [pinia, i18n],
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
    vi.clearAllMocks();
  });

  describe('Component rendering', () => {
    it('should match snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders the modal with correct props', () => {
      const modal = findComponent('modal');
      expect(modal.exists()).toBe(true);
      expect(modal.props('modelValue')).toBe(true);
      expect(modal.props('title')).toBe(
        i18n.global.t(
          'agent_builder.instructions.new_instruction.validate_instruction_by_ai.modal_title',
        ),
      );
      expect(modal.props('showCloseIcon')).toBe(true);
      expect(modal.props('showActionsDivider')).toBe(true);
      expect(modal.props('size')).toBe('lg');
    });

    it('renders the correct button props', () => {
      const modal = findComponent('modal');
      expect(modal.props('secondaryButtonProps').text).toBe(
        i18n.global.t(
          'agent_builder.instructions.new_instruction.validate_instruction_by_ai.cancel_button',
        ),
      );

      expect(modal.props('primaryButtonProps').text).toBe(
        i18n.global.t(
          'agent_builder.instructions.new_instruction.validate_instruction_by_ai.publish_button',
        ),
      );
      expect(modal.props('primaryButtonProps').disabled).toBe(false);
      expect(modal.props('primaryButtonProps').loading).toBe(false);
    });

    it('renders the instruction textarea with correct props', () => {
      const instructionTextarea = findComponent('instructionTextarea');
      expect(instructionTextarea.exists()).toBe(true);
      expect(instructionTextarea.props('modelValue')).toBe(
        instructionsStore.newInstruction.text,
      );
    });

    describe('Validation results rendering', () => {
      it('renders the instruction divider', () => {
        const instructionDivider = findComponent('instructionDivider');
        expect(instructionDivider.exists()).toBe(true);
      });

      it('renders the instruction validation results with correct props', () => {
        const instructionValidationResults = findComponent(
          'instructionValidationResults',
        );
        expect(instructionValidationResults.exists()).toBe(true);
      });

      it('does not render the instruction validation results when the instruction suggested by AI status is error', async () => {
        instructionsStore.instructionSuggestedByAI.status = 'error';
        await wrapper.vm.$nextTick();
        expect(findComponent('instructionValidationResults').exists()).toBe(
          false,
        );
      });
    });
  });

  describe('Modal interactions', () => {
    it('emits update:modelValue when modal is closed via close icon', async () => {
      await findComponent('modal').vm.$emit('update:model-value', false);

      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
      expect(wrapper.emitted('update:modelValue')[0]).toEqual([false]);
    });

    it('emits update:modelValue when secondary button is clicked', async () => {
      const modal = findComponent('modal');
      await modal.props('secondaryButtonProps').onClick();

      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
      expect(wrapper.emitted('update:modelValue')[0]).toEqual([false]);
    });
  });

  describe('Publish instruction functionality', () => {
    it('calls addInstruction when primary button is clicked', async () => {
      const modal = findComponent('modal');
      await modal.props('primaryButtonProps').onClick();

      expect(instructionsStore.addInstruction).toHaveBeenCalled();
    });

    it('updates newInstruction.text before calling addInstruction', async () => {
      const modal = findComponent('modal');
      const newText = 'Updated instruction text';
      wrapper.vm.newInstruction = newText;

      await modal.props('primaryButtonProps').onClick();

      expect(instructionsStore.newInstruction.text).toBe(newText);
      expect(instructionsStore.addInstruction).toHaveBeenCalled();
    });
  });

  describe('Button disabled states', () => {
    it('disables primary button when instructionSuggestedByAI status is loading', async () => {
      instructionsStore.instructionSuggestedByAI.status = 'loading';
      await wrapper.vm.$nextTick();

      const modal = findComponent('modal');
      expect(modal.props('primaryButtonProps').disabled).toBe(true);
    });

    it('disables primary button when newInstruction is empty', async () => {
      wrapper.vm.newInstruction = '';
      await wrapper.vm.$nextTick();

      const modal = findComponent('modal');
      expect(modal.props('primaryButtonProps').disabled).toBe(true);
    });

    it('disables primary button when newInstruction is only whitespace', async () => {
      wrapper.vm.newInstruction = '   ';
      await wrapper.vm.$nextTick();

      const modal = findComponent('modal');
      expect(modal.props('primaryButtonProps').disabled).toBe(true);
    });

    it('enables primary button when newInstruction has valid text', async () => {
      instructionsStore.instructionSuggestedByAI.status = null;
      wrapper.vm.newInstruction = 'Valid instruction text';
      await wrapper.vm.$nextTick();

      const modal = findComponent('modal');
      expect(modal.props('primaryButtonProps').disabled).toBe(false);
    });
  });

  describe('Button loading states', () => {
    it('shows loading state on primary button when newInstruction status is loading', async () => {
      instructionsStore.newInstruction.status = 'loading';
      await wrapper.vm.$nextTick();

      const modal = findComponent('modal');
      expect(modal.props('primaryButtonProps').loading).toBe(true);
    });

    it('does not show loading state when newInstruction status is not loading', async () => {
      instructionsStore.newInstruction.status = null;
      await wrapper.vm.$nextTick();

      const modal = findComponent('modal');
      expect(modal.props('primaryButtonProps').loading).toBe(false);
    });
  });

  describe('onMounted behavior', () => {
    it('calls getInstructionSuggestionByAI when instruction is not suggested', () => {
      expect(instructionsStore.getInstructionSuggestionByAI).toHaveBeenCalled();
    });

    it('does not call getInstructionSuggestionByAI when instruction is already suggested', () => {
      vi.clearAllMocks();

      instructionsStore.instructionSuggestedByAI.data.instruction =
        'Test instruction text';
      instructionsStore.newInstruction.text = 'Test instruction text';

      wrapper = mount(ModalValidateInstruction, {
        props: {
          modelValue: true,
        },
      });

      expect(
        instructionsStore.getInstructionSuggestionByAI,
      ).not.toHaveBeenCalled();
    });
  });
});
