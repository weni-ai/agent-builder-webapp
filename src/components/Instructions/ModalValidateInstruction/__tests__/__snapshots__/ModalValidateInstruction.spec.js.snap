// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModalValidateInstruction.vue > Component rendering > should match snapshot 1`] = `
"<section data-v-ab1febba="" class="unnnic-modal-dialog" data-testid="modal-validate-instruction-by-ai">
  <section data-v-ab1febba="" class="unnnic-modal-dialog__overlay" data-testid="modal-overlay"></section>
  <section data-v-ab1febba="" class="unnnic-modal-dialog__container unnnic-modal-dialog__container--lg" data-testid="modal-container">
    <!---->
    <section data-v-ab1febba="" class="unnnic-modal-dialog__container__body">
      <header data-v-ab1febba="" class="unnnic-modal-dialog__container__header">
        <section data-v-ab1febba="" class="unnnic-modal-dialog__container__title-container">
          <!---->
          <h1 data-v-ab1febba="" class="unnnic-modal-dialog__container__title-text" data-testid="title-text">Validate instruction</h1>
        </section><span data-v-fb16c4d6="" data-v-ab1febba="" class="material-symbols-rounded unnnic-icon-scheme--neutral-darkest unnnic-icon-size--md unnnic--clickable" data-testid="close-icon" translate="no">close</span>
      </header>
      <section data-v-ab1febba="" class="unnnic-modal-dialog__container__content"></section>
      <section data-v-ab1febba="" data-testid="actions-section" class="unnnic-modal-dialog__container__actions unnnic-modal-dialog__container__actions--divider"><button data-v-992d21f0="" data-v-ab1febba="" data-testid="secondary-button" class="unnnic-modal-dialog__container__actions__secondary-button unnnic-button unnnic-button--size-large unnnic-button--tertiary unnnic-modal-dialog__container__actions__secondary-button">
          <!---->
          <!---->
          <!----><span data-v-992d21f0="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> Cancel</span>
          <!---->
        </button><button data-v-992d21f0="" data-v-ab1febba="" data-testid="primary-button" class="unnnic-modal-dialog__container__actions__primary-button unnnic-button unnnic-button--size-large unnnic-button--primary unnnic-modal-dialog__container__actions__primary-button">
          <!---->
          <!---->
          <!----><span data-v-992d21f0="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> Publish</span>
          <!---->
        </button></section>
    </section>
  </section>
</section>"
`;
